egauge/script/api_egauge.py

```diff
@@ -0,0 +1,148 @@
+                                # the upload timestamp for all rows in the dataframe will use the same value (current_time) which was set at the beginning of pull_egauge_data()
+                                reading_row = orm_egauge.Reading(sensor_id=sensor_id, timestamp=row_datetime, units=row_units, reading=row_reading, upload_timestamp=current_time.datetime)
+                                session.add(reading_row)
+                                last_inserted_time = arrow.get(row_datetime)
+                        is_success = True
+                    request_timer_end = arrow.now()
+                    request_time_elapsed = request_timer_end - request_timer_start
+                    print(str(rows_returned) + ' row(s) returned by egauge api in ' + str(request_time_elapsed))
+                else:
+                    request.raise_for_status()
+        #catch egauge api request exceptions like requests.exceptions.ConnectionError
+        except Exception as e:
+            error_msg = str(traceback.format_exc())
+            print(error_msg)
+            with open(ERROR_LOG, 'a+') as error_file:
+                error_time_readable = arrow.now().format('ddd MMM DD YYYY HH:mm:ss ZZ')
+                error_file.write('[' + error_time_readable + ']')
+                error_file.write(error_msg + '\n')
+        # if statement accounts cases where the egauage sensor clock is behind the server clock
+        # guarantees that the next api run will use a start time one minute after the last inserted reading time if there is at least one success in database_insertion_timestamp
+        if last_inserted_time:
+            datetime_to_insert = last_inserted_time.shift(minutes = +1).datetime
+        else:
+            datetime_to_insert = current_time.datetime
+        timestamp_row = orm_egauge.DatabaseInsertionTimestamp(timestamp=datetime_to_insert, is_success=is_success)
+        session.add(timestamp_row)
+        # #TESTING
+        # database_insertion_timestamps = session.query(orm_egauge.DatabaseInsertionTimestamp)
+        # for database_insertion_timestamp in database_insertion_timestamps:
+        #     print("values in database_insertion_timestamp are: " + str(database_insertion_timestamp.timestamp) + ', ' + str(database_insertion_timestamp.is_success))
+        session.commit()
+        session.close()
+    #catch ORM exeptions like sqlalchemy.exc.InternalError, sqlalchemy.exc.OperationalError
+    except Exception as e:
+        error_msg = str(traceback.format_exc())
+        print(error_msg)
+        with open(ERROR_LOG, 'a+') as error_file:
+            error_time_readable = arrow.now().format('ddd MMM DD YYYY HH:mm:ss ZZ')
+            error_file.write('[' + error_time_readable + ']')
+            error_file.write(error_msg + '\n')
+
+if __name__ == "__main__":
+    # slice off script name argument since it is unused
+    pull_egauge_data(*sys.argv[1:])
```

Comment:
We need to make the main look like a main, i.e. I need to understand the major steps in the code flow by reading this at a higher abstract level now that we have the code start to end.

If we were to talk about this out loud I would think of these steps:
 
 * last_reading_timestamp = get_most_recent_timestamp_from_db(db_config1,db_config2,...,db_config_n)
 * try: 
    * readings,api_call_timestamp = get_data_from_api(last_reading_timestamp) # readings is a pandas dataframe. 
 * except: 
    * log_failure_to_connect_to_api(api_call_timestamp) 
    * exit 
 * try: 
    * database_insertion_timestamp = insert_readings_into_database(readings) 
 * except: 
   * log_failure_to_connect_to_database(database_insertion_timestamp)

Reply:
Also take note why I insisted on you not creating all those functions from that single notebook early on: All of them, and the notebook itself is still a single function and code flow in this definition (get_data_from_api). 

Note also at this level of abstraction, the "main" function applies exactly to both webctrl and egauge.

If we want

Reply:
For the function __get_data_from_api()__ what do you think if the function definition with parameters looks something like this?

__get_data_from_api(api_start_time, api_end_time, sensor_id, unit_of_time):__

A call to that function could look something like this:

__readings, api_call_time = get_data_from_api(last_reading_time, api_current_time, sensor_id, unit_of_time)__

That way it can also be used to test an egauge api call with an end time different from the current time.

Note: __unit_of_time__ is used in the egauge api request and is usually set to the string 'm' representing minute.